dApps staking v3 - part 1

Dinonard opened the pull request:
**Pull Request Summary**

First PR that will introduce part of dApps staking v3 functionality.
Overall implementation will be broken down into multiple PRs to make it easier for reviewers.

**Part 1**
* maintenance mode
* basic operations for registering dApps
* dApp unregistration, reward destination & owner manipulation
* various struct & types to simplify the implementation
* balance lock implementation

After receiving first review, will amend the comments and implement extensive unit tests.

**Check list**
- [x] dApp manipulation & balance lock implemented 
- [x] first review received
- [x] added unit tests

Dinonard committed:
dApps staking v3 - inception

Dinonard committed:
First ext call wip

Dinonard committed:
Register & reward_dest implemented

Dinonard committed:
Reward beneficiary & owner changes

Dinonard committed:
Remove register deposit

Dinonard committed:
UNregister & various minor improvements

Dinonard committed:
Progress towards lock

Dinonard committed:
Lock progress

Dinonard committed:
Lock progress

Dinonard committed:
Lock implemented & some code improvements

gitofdeepanshu commented on frame/dapp-staking-v3/src/lib.rs:
```diff
@@ -0,0 +1,693 @@
+                },
+            )?;
+
+            // TODO: might require some modification later on, like additional checks to ensure contract can be unregistered.
```
what are the scenarios in which contract can't be unregistered?

Dinonard replied:
What should be avoided is unregistering a contract while leaving storage items behind, _forever_.
So later on, when PR progresses more, this TODO is a reminder to revisit the issue (I don't have a clear solution for this yet).

gitofdeepanshu commented on frame/dapp-staking-v3/src/lib.rs:
```diff
@@ -0,0 +1,693 @@
+    #[codec(compact)]
+    amount: Balance,
+    #[codec(compact)]
+    unlock_block: BlockNumber,
```
`BlockNumber` is undefined here.

Dinonard replied:
Right, late minute breaking changes 😂

gitofdeepanshu commented on frame/dapp-staking-v3/src/lib.rs:
```diff
@@ -0,0 +1,693 @@
+pub type BalanceOf<T> =
+    <<T as Config>::Currency as Currency<<T as frame_system::Config>::AccountId>>::Balance;
+
+/// Concenience type for `AccountLedger` usage.
```
is this a typo? `convenient` maybe?

Dinonard replied:
Yes, should be convenience.

Dinonard committed:
PR comments

Dinonard committed:
Tests & some changes

Dinonard committed:
Additional UTs

shaunxw commented on frame/dapp-staking-v3/src/lib.rs:
```diff
@@ -0,0 +1,730 @@
+#[cfg(test)]
+mod test;
+
+const STAKING_ID: LockIdentifier = *b"dapstake";
```
I wonder what's the migration plan to v3? If v3 coexists with prev version, probably best to use a different lock ID.

Dinonard replied:
v3 will replace the old version, and we want to keep the existing locks.

shaunxw replied:
I see. Then it makes sense to use the existing lock ID.

shaunxw commented on frame/dapp-staking-v3/src/lib.rs:
```diff
@@ -0,0 +1,730 @@
+pub enum PeriodType {
+    /// Period during which the focus is on voting.
+    /// Inner value is the era in which the voting period ends.
+    VotingPeriod(#[codec(compact)] EraNumber),
```
```suggestion
    Voting {
        #[codec(compact)]
        ends_at: EraNumber,
    }),
```

It is more about personal preference, but I think this is more readable and self-explanatory. `PeriodType` naming includes `Period` space so we don't need to repeat.

If it makes sense, we can apply this to other enum fields.

Dinonard replied:
I agree 😁 

Haven't gotten to the part of implementation where I use this, but I'd probably get the same idea as you after writing down `PeriodType::VotingPeriod` xD

shaunxw commented on frame/dapp-staking-v3/src/lib.rs:
```diff
@@ -0,0 +1,730 @@
+    }
+}
+
+/// Rewards pool for lock participants  & dApps
```
```suggestion
/// Rewards pool for lock participants & dApps
```

shaunxw commented on frame/dapp-staking-v3/src/lib.rs:
```diff
@@ -0,0 +1,730 @@
+            dapp_id: DAppId,
+        },
+        /// dApp reward destination has been updated.
+        DAppRewardDestination {
```
```suggestion
        DAppRewardDestinationUpdated {
```

shaunxw commented on frame/dapp-staking-v3/src/lib.rs:
```diff
@@ -0,0 +1,730 @@
+                .map_err(|_| Error::<T>::TooManyLockedBalanceChunks)?;
+            ensure!(
+                ledger.locked_amount().saturating_add(amount_to_lock)
+                    > T::MinimumLockedAmount::get(),
```
```suggestion
                    >= T::MinimumLockedAmount::get(),
```

A minimum amout should be valid right?

Dinonard replied:
Right, also discovered this via UT 🙂

shaunxw reviewed (commented):
Looks good in general!

I wonder why we are using `DispatchResultWithPostInfo` as the return type though there is no post info?

shaunxw commented on frame/dapp-staking-v3/src/lib.rs:
```diff
@@ -0,0 +1,730 @@
+    pub owner: AccountId,
+    /// dApp's unique identifier in dApp staking.
+    #[codec(compact)]
+    pub id: DAppId,
```
Forgot to add this one: do we need `id` when they can be identified via `T::SmartContract`?

Dinonard replied:
ID was added because of how dApp rewards will be calculated.
Using `T::SmartContract` as identifier works, logically but it takes a lot of space.
To optimize this, a more compact _secundary_ (not sure if that's the right term) is introduced.

Right now it's part of the design (more info in the doc) but it might change in the future.

Dinonard committed:
Lock tests

Dinonard committed:
PR comments

Dinonard commented:
> Looks good in general!
> 
> I wonder why we are using `DispatchResultWithPostInfo` as the return type though there is no post info?

Thanks!

For the `DispatchResultWithPostInfo` you're right that it's not needed. I just copied it from the previous pallet for the first extrinsic I added, then reused it afterwards 😁. Fixed it now.

Dinonard commented:
All functionality intended for this PR has been implemented, and it's ready for final review 🙏

Dinonard committed:
Introduce types

shaunxw commented on frame/dapp-staking-v3/src/lib.rs:
```diff
@@ -0,0 +1,450 @@
+        /// Smart contract already exists within dApp staking protocol.
+        ContractAlreadyExists,
+        /// Maximum number of smart contracts has been reached.
+        ExcededMaxNumberOfContracts,
```
```suggestion
        ExceededMaxNumberOfContracts,
```

shaunxw commented on frame/dapp-staking-v3/src/lib.rs:
```diff
@@ -0,0 +1,450 @@
+                dapp_id,
+            });
+
+            Ok(().into())
```
```suggestion
            Ok(())
```

`into` isn't needed anymore with `DispatchResult` as return type. Same to other calls.

Dinonard replied:
Right, I rushed! Will clean it up!

shaunxw commented on frame/dapp-staking-v3/src/test/mock.rs:
```diff
@@ -0,0 +1,182 @@
+
+/// Initialize first block.
+/// This method should only be called once in a UT otherwise the first block will get initialized multiple times.
+pub(crate) fn initialize_first_block() {
```
Could we have it executed in `build` function, so it doesn't have to repeat in each UT?

Dinonard replied:
Sure, good suggestion! Nice way to get `before_each` hook 🙂

shaunxw commented on frame/dapp-staking-v3/src/lib.rs:
```diff
@@ -0,0 +1,450 @@
+    #[pallet::error]
+    pub enum Error<T> {
+        /// Pallet is disabled/in maintenance mode.
+        Disabled,
```
IMO disabled/maintainance mode should also be covered in unit tests.

Dinonard replied:
True, I didn't forget about it though, just thought I'd cover it at the very end.
Personal bias since for `v2`, I added maintenance mode only later.

Will add it now though!

shaunxw reviewed (commented):
Looks good overall! Just some small issues.

Dinonard committed:
Update frame/dapp-staking-v3/src/lib.rs

Co-authored-by: Shaun Wang <spxwang@gmail.com>

Dinonard committed:
Address PR review comments

Dinonard commented:
@shaunxw thanks for the approval!

I was thinking, instead of merging it directly, to keep this in a 'live' branch.
All consecutive PRs would be merged to this branch instead (or I could create another branch to aggregate all PRs).

This would keep the review of PRs light(er), while also providing one big overview in the end, before everything is merged.
What do you think?

shaunxw commented:
Love this idea. It definitely helps with the review process.
